compiler/rustc_mir_dataflow/src/impls/live_borrows.rs

```diff
@@ -0,0 +1,848 @@
+    AnalysisDomain, Backward, CallReturnPlaces, GenKill, GenKillAnalysis, ResultsRefCursor,
+};
+use rustc_data_structures::fx::FxHashMap;
+use rustc_data_structures::graph::implementation::{Graph, NodeIndex};
+use rustc_middle::mir::visit::PlaceContext;
+use rustc_middle::mir::visit::Visitor;
+use rustc_middle::mir::*;
+use rustc_middle::ty;
+
+use either::Either;
+
+#[derive(Copy, Clone, Debug)]
+enum NodeKind {
+    // An node corresponding to the place on the lhs of an assignment like `_3 = Ref(_, _, _4)`.
+    Borrow(Local),
+
+    // Nodes corresponding to the place on the lhs of a statement like
+    // `_2 = Aggregate(Adt(..), _, _, _, _), [move _3, move _6])`,
+    // where _3 and _6 are places corresponding to references or raw pointers.
+    LocalWithRefs(Local),
+
+    // Nodes corresponding to the borrowed place of an assignment like `_2 = Ref(_, _, borrowed_place)`,
+    // if `borrowed_place` is a non-ref or non-ptr value.
+    Local(Local),
+}
+
+impl NodeKind {
+    fn get_local(&self) -> Local {
+        match self {
+            Self::Borrow(local) => *local,
+            Self::LocalWithRefs(local) => *local,
+            Self::Local(local) => *local,
+        }
+    }
+}
+
+/// Used to build a dependency graph between borrows/pointers and the `Local`s that
+/// they reference.
+/// We add edges to the graph in following situations:
+///     * direct assignment of reference or raw pointer (e.g. `_4 = Ref(_, _ , borrowed_place)` or
+///       `_4 = AddressOf(_, borrowed_place)`). For this case we create a `Node` of kind
+///       `NodeKind::Borrow` for the `Local` being assigned to and an edge to either an existing
+///       `Node` or if none exists yet to a new `Node` of type `NodeKind::Local` corresponding to
+///       a non-ref/ptr `Local`.
+///     * assignments to non-reference or non-pointer `Local`s, which themselves might contain
+///       references or pointers (e.g. `_2 = Aggregate(Adt(..), _, _, _, _), [move _3, move _6])`,
+///       where `_3` and `_6` are places corresponding to references or raw pointers). In this case
+///       we create a `Node` of kind `NodeKind::LocalWithRefs` for `_2`. Since `_3` and `_6` are
+///       `Local`s that correspond to references, pointers or composite types that might contain
+///       references or pointers (`NodeKind::LocalWithRefs`), there already exist `Node`s for these
+///       `Local`s. We then add edges from the `Node` for `_2` to both the `Node` for `_3` and the
+///       `Node` for `_6`.
+///     * `destination` places for `TerminatorKind::Call` and the `resume_arg` places for
+///        `TerminatorKind::Yield` if we pass in any references, pointers or composite values that
+///        might correspond to references, pointers or exposed pointers (`NodeKind::LocalWithRef`s).
+///        The rationale for this is that the return values of both of these terminators might themselves
+///        contain any of the references or pointers passed as arguments.
```

Comment:
This documentation is explaining the *algorithm* that is implemented. That's not the most important part though. The `BorrowDependencies` struct and its associated logic first and foremost represents an *analysis* of a mir body. The documentation needs to describe what the results from that analysis are. In other words, this currently says something like "this analysis works by doing steps 1, 2, and 3." I'm looking for something that says, "if the result of this analysis is a graph that looks like G, then you can conclude that at location L in the body, thing X is true" (where this'll be something about liveness or whatever).

Reply:
`BorrowDependencies` is only responsible for constructing the dependency graph. I tried to more elaborate what results the analysis as a whole produces in the doc comment. I still think that the comment for `BorrowDependencies` is fairly adequate given that it's a private type and only used in the live borrows analysis.